### DISCUSSION TITLE
Build error with pyrfc 2.4.6

### DISCUSSION THREAD

**Discussion by Kobablood** [SAP/PyRFC#247] [Q&A] [answered]
I want install pyrfc in my windows 10 os.
My python version is 3.9.6,so i download the sap nwrfcsdk from smp,version is 750 Patch Level 8,then download the last source of pyrfc version is 2.4.2.
Before the build wheel，i have installed vs build tool（2005,2013，2016 and 2019).

But build still failed,the error as below:
 
```shell
python setup.py bdist_wheel

Output:

750 Patch Level 8
running bdist_wheel
running build
running build_py
creating build
creating build\lib.win32-3.9
creating build\lib.win32-3.9\pyrfc
copying src\pyrfc\_exception.py -> build\lib.win32-3.9\pyrfc
copying src\pyrfc\__init__.py -> build\lib.win32-3.9\pyrfc
running build_ext
skipping 'src/pyrfc\_pyrfc.cpp' Cython extension (up-to-date)
building 'pyrfc.pyrfc' extension
creating build\temp.win32-3.9
creating build\temp.win32-3.9\Release\src
creating build\temp.win32-3.9\Release\src\pyrfc
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX86\x86\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -DSAPonNT -D_CRT_NON_CONFORMING_SWPRINTFS -D_CRT_SECURE_NO_DEPRECATES -D_CRT_NONSTDC_NO_DEPRECATE -D_AFXDLL -DWIN32 -D_WIN32_WINNT=0x0502 -DWIN64 -D_AMD64_ -DNDEBUG -DSAPwithUNICODE -DUNICODE -D_UNICODE -DSAPwithTHREADS -D_ATL_ALLOW_CHAR_UNSIGNED -D_LARGEFILE_SOURCE -D_CONSOLE -DSAP_PLATFORM_MAKENAME=ntintel -IF:\python3\include -IF:\python3\include -IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\ATLMFC\include -IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt /EHsc /Tpsrc/pyrfc\_pyrfc.cpp /Fobuild\temp.win32-3.9\Release\src/pyrfc\_pyrfc.obj -IF:\nwrfcsdk\include -IF:\python3\lib\inspect.py\Include -IF:\python3\lib\inspect.py\Include\PC /EHs /Gy /J /MD /nologo /W3 /Z7 /GL /O2 /Oy- /we4552 /we4700 /we4789
_pyrfc.cpp
F:\nwrfcsdk\include\sapnwrfc.h(5404): warning C4819: 该文件包含不能在当前代码页(936)中表示的字符。请将该文件保存为 Unicode 格式以防止数据丢失       
src/pyrfc\_pyrfc.cpp(21148): error C2664: “RFC_RC RfcInstallGenericServerFunction(RFC_SERVER_FUNCTION,RFC_FUNC_DESC_CALLBACK,RFC_ERROR_INFO *)”: 无 
法将参数 1 从“RFC_RC (__cdecl *)(RFC_CONNECTION_HANDLE,RFC_FUNCTION_HANDLE,RFC_ERROR_INFO *)”转换为“RFC_SERVER_FUNCTION”
src/pyrfc\_pyrfc.cpp(21148): note: 在匹配目标类型的范围内没有具有该名称的函数
F:\nwrfcsdk\include\sapnwrfc.h(2266): note: 参见“RfcInstallGenericServerFunction”的声明
src/pyrfc\_pyrfc.cpp(29685): warning C4018: “<”: 有符号/无符号不匹配
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x86\\cl.exe' failed with exit code 2
```
I notice the core error is "error C2664: “RFC_RC RfcInstallGenericServerFunction(RFC_SERVER_FUNCTION,RFC_FUNC_DESC_CALLBACK,RFC_ERROR_INFO *)”: 无 
法将参数 1 从“RFC_RC (__cdecl *)(RFC_CONNECTION_HANDLE,RFC_FUNCTION_HANDLE,RFC_ERROR_INFO *)”转换为“RFC_SERVER_FUNCTION”"

How can I install the new version pyrfc,may you help me.

Thanks.

Vincent Chen

**Comment by bsrdjan** [comment, marked as answer]
You are trying to build the 32-bit PyRFC correct?

Could you please let me know:
 
- The Windows machine is 32 or 64 bit?
- Python 3.9.5 32 or 64 bit?
- SAP NWRFC SDK is 32 or 64 bit?
- VS C++ redistributable 2013 installed ? 32 or 64 bit? 

All prerequisites are listed here: https://github.com/SAP/PyRFC#windows

**Reply by Kobablood** [reply]
Dear Bsrdjan,

I want to build the 32-bit Pyrfc.

The Windows machine is 32 or 64 bit?           
A:64Bit
Python 3.9.6 32 or 64 bit?                                 
A:('32bit', 'WindowsPE')

SAP NWRFC SDK is 32 or 64 bit?                      

A: I have not know SAP NWRFC SDK  is 32bit or 64bit because only one version dowload from SMP,i think this is 64bit.

VS C++ redistributable 2013 installed ? 32 or 64 bit? 

A: 32 and 64 bit all installed.

Maybe should I try to install 64 bit python for build？

Vincent Chen

**Comment by bsrdjan** [comment]
32-bit Python. is fine and you need 32 bit RFC SDK

The 32-bit RFC SDK can be downloaded from the link in [SAP OSS Note 2573790 - Installation, Support and Availability of the SAP NetWeaver RFC Library 7.50](https://launchpad.support.sap.com/#/notes/2573790).

It is called `Windows Server`

<img width="791" alt="Screenshot 2021-10-14 at 10 41 05" src="https://user-images.githubusercontent.com/1697847/137282991-1a8daad4-5ae5-4bbc-a6af-19b229d493f4.png">

**Comment by Kobablood** [comment]
Dear bsrdjan,

I have reinstalled 64-it python 3.9.7,Everything is ok,no build error.
Thanks very much..

Vincent Chen
